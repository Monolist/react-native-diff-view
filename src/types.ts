import { IHunk, IChange } from 'gitdiff-parser';
import { TextStyle, ViewStyle } from 'react-native';

export type DiffType = 'add' | 'delete' | 'modify' | 'rename' | 'copy';

export type Side = 'old' | 'new';

export type ViewType = 'unified' | 'split';

export type GutterType = 'default' | 'none' | 'anchor';

export type Widgets = Record<string, JSX.Element>;

export type Events = Record<string, (arg: any) => any>;

export type DiffChildren = (hunks: IHunk[]) => JSX.Element | JSX.Element[];

export interface IGutterOptions {
  change: IChange;
  side: Side;
  inHoverState: boolean;
  textStyle?: TextStyle;
  renderDefault: (style?: TextStyle, textStyle?: ViewStyle) => JSX.Element;
}
