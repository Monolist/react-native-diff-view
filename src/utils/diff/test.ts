import { IHunk, IChange } from '../../types';
import { basicHunk, normalChange, insertChange, deleteChange } from '../../fixtures';
import {
  computeLineNumberFactory,
  isInHunkFactory,
  isBetweenHunksFactory,
  findChangeByLineNumberFactory,
  getCorrespondingLineNumberFactory,
} from '../diff/factory';
import {
  textLinesToHunk,
  insertHunk,
  expandFromRawCode,
  getCollapsedLinesCountBetween,
  expandCollapsedBlockBy,
  getChangeKey,
} from '..';

describe('utils/diff', () => {
  describe('textLinesToHunk', () => {
    test('basic', () => {
      const lines = [''];

      expect(textLinesToHunk(lines, 0, 0)).toMatchSnapshot();
    });
  });

  describe('insertHunk', () => {
    test('basic', () => {
      expect(
        insertHunk([{ changes: [{ isInsert: true, lineNumber: 0, oldLineNumber: 0, type: 'insert' }] } as IHunk], {
          changes: [{ isInsert: true, lineNumber: 1, oldLineNumber: 1, type: 'insert' }],
        } as IHunk),
      ).toEqual([{ changes: [{ isInsert: true, lineNumber: 1, oldLineNumber: 1, type: 'insert' }] }]);
    });
  });

  describe('expandFromRawCode', () => {
    test('basic', () => {
      const hunks = [basicHunk];

      expect(expandFromRawCode(hunks, '')).toMatchSnapshot();
    });
  });

  describe('getCollapsedLinesCountBetween', () => {
    test('empty', () => {
      const previousHunk = {} as IHunk;
      const nextHunk = {} as IHunk;

      expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(NaN);
    });

    test('basic', () => {
      const previousHunk = { oldStart: 1, oldLines: 2 } as IHunk;
      const nextHunk = { oldStart: 10 } as IHunk;

      expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(7);
    });

    test('minus number', () => {
      const previousHunk = { oldStart: 1, oldLines: 10 } as IHunk;
      const nextHunk = { oldStart: 2 } as IHunk;

      expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(-9);
    });

    test('no previousHunk', () => {
      const nextHunk = { oldStart: 2 } as IHunk;

      expect(getCollapsedLinesCountBetween(null, nextHunk)).toBe(1);
    });

    test('throw when nextHunk is null', () => {
      expect(() => getCollapsedLinesCountBetween({} as IHunk, null)).toThrow();
    });
  });

  describe('expandCollapsedBlockBy', () => {
    test('basic', () => {
      const hunks = [basicHunk];
      const noop = (): boolean => true;

      expect(expandCollapsedBlockBy(hunks, '', noop)).toMatchSnapshot();
    });
  });

  describe('getChangeKey', () => {
    test('normal change', () => {
      expect(getChangeKey(normalChange)).toBe('N0');
    });

    test('insert change', () => {
      expect(getChangeKey(insertChange)).toBe('I0');
    });

    test('delete change', () => {
      expect(getChangeKey(deleteChange)).toBe('D0');
    });
  });

  describe('factory', () => {
    test('returns a function', () => {
      expect(typeof (computeLineNumberFactory as any)()).toBe('function');
      expect(typeof (isInHunkFactory as any)()).toBe('function');
      expect(typeof (isBetweenHunksFactory as any)()).toBe('function');
      expect(typeof (findChangeByLineNumberFactory as any)()).toBe('function');
      expect(typeof (getCorrespondingLineNumberFactory as any)()).toBe('function');
    });
  });

  describe('computeLineNumber', () => {
    test('old', () => {
      const computeOldLineNumber = computeLineNumberFactory('old');

      expect(computeOldLineNumber({ isInsert: true } as IChange)).toBe(-1);
      expect(computeOldLineNumber({ isNormal: true, lineNumber: 0, oldLineNumber: 1 } as IChange)).toBe(1);
      expect(computeOldLineNumber({ isNormal: false, lineNumber: 0, oldLineNumber: 1 } as IChange)).toBe(0);
    });

    test('new', () => {
      const computeNewLineNumber = computeLineNumberFactory('new');

      expect(computeNewLineNumber({ isDelete: true } as IChange)).toBe(-1);
      expect(computeNewLineNumber({ isNormal: true, lineNumber: 0, newLineNumber: 1 } as IChange)).toBe(1);
      expect(computeNewLineNumber({ isNormal: false, lineNumber: 0, newLineNumber: 1 } as IChange)).toBe(0);
    });
  });

  describe('isInHunk', () => {
    test('old', () => {
      const isInOldHunk = isInHunkFactory('oldStart', 'oldLines');

      expect(isInOldHunk({ oldStart: 1, oldLines: 2 } as IHunk, 2)).toBe(true);
      expect(isInOldHunk({ oldStart: 1, oldLines: 2 } as IHunk, 3)).toBe(false);
    });

    test('new', () => {
      const isInNewHunk = isInHunkFactory('newStart', 'newLines');

      expect(isInNewHunk({ newStart: 1, newLines: 2 } as IHunk, 2)).toBe(true);
      expect(isInNewHunk({ newStart: 1, newLines: 2 } as IHunk, 3)).toBe(false);
    });
  });

  describe('isBetweenHunks', () => {
    test('old', () => {
      const isOldLineNumberBetweenHunks = isBetweenHunksFactory('oldStart', 'oldLines');

      expect(isOldLineNumberBetweenHunks({ oldStart: 1, oldLines: 2 } as IHunk, { oldStart: 4 } as IHunk, 2)).toBe(
        false,
      );
      expect(isOldLineNumberBetweenHunks({ oldStart: 1, oldLines: 2 } as IHunk, { oldStart: 4 } as IHunk, 3)).toBe(
        true,
      );
    });

    test('new', () => {
      const isNewLineNumberBetweenHunks = isBetweenHunksFactory('newStart', 'newLines');

      expect(isNewLineNumberBetweenHunks({ newStart: 1, newLines: 2 } as IHunk, { newStart: 4 } as IHunk, 2)).toBe(
        false,
      );
      expect(isNewLineNumberBetweenHunks({ newStart: 1, newLines: 2 } as IHunk, { newStart: 4 } as IHunk, 3)).toBe(
        true,
      );
    });
  });

  describe('findChangeByLineNumber', () => {
    test('old', () => {
      const findChangeByLineNumber = findChangeByLineNumberFactory('old');
      const change = { isNormal: true, lineNumber: 0, oldLineNumber: 1 } as IChange;
      const hunk = { oldStart: 1, oldLines: 2, changes: [change] } as IHunk;

      expect(findChangeByLineNumber([hunk], 1)).toBe(change);
      expect(findChangeByLineNumber([hunk], 3)).toBe(undefined);
    });

    test('new', () => {
      const findChangeByLineNumber = findChangeByLineNumberFactory('new');
      const change = { isNormal: true, lineNumber: 0, newLineNumber: 1 } as IChange;
      const hunk = { newStart: 1, newLines: 2, changes: [change] } as IHunk;

      expect(findChangeByLineNumber([hunk], 1)).toBe(change);
      expect(findChangeByLineNumber([hunk], 3)).toBe(undefined);
    });
  });

  describe('getCorrespondingLineNumber', () => {
    test('old', () => {
      // getNewCorrespondingLineNumber is the same
      const getOldCorrespondingLineNumber = getCorrespondingLineNumberFactory('old');
      const hunk = { oldStart: 10, oldLines: 5, newStart: 20, newLines: 5, changes: [] } as IHunk;

      expect(() => getOldCorrespondingLineNumber([], 0)).toThrow();
      expect(getOldCorrespondingLineNumber([hunk], 0)).toBe(10);
      expect(getOldCorrespondingLineNumber([hunk], 20)).toBe(30);

      hunk.changes = [{ isNormal: true, oldLineNumber: 11 } as IChange];
      expect(getOldCorrespondingLineNumber([hunk], 11)).toBe(undefined);

      hunk.changes = [{ isNormal: true, oldLineNumber: 12, newLineNumber: 22 } as IChange];
      expect(getOldCorrespondingLineNumber([hunk], 12)).toBe(22);

      hunk.changes = [{ isInsert: true, lineNumber: 13 } as IChange];
      expect(() => getOldCorrespondingLineNumber([hunk], 13)).toThrow();

      hunk.changes = [{ isDelete: true, lineNumber: 14 } as IChange];
      expect(getOldCorrespondingLineNumber([hunk], 14)).toBe(-1);

      const nextHunk = { oldStart: 20, oldLines: 5, newStart: 30, newLines: 5, changes: [] } as IHunk;

      expect(getOldCorrespondingLineNumber([hunk, nextHunk], 16)).toBe(26);
    });
  });
});
