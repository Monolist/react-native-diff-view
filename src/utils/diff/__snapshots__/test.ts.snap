// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/diff expandCollapsedBlockBy basic 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "",
        "isNormal": true,
        "newLineNumber": 1,
        "oldLineNumber": 1,
        "type": "normal",
      },
    ],
    "content": "@@ -1,1 +1,1",
    "isPlain": true,
    "newLines": 1,
    "newStart": 1,
    "oldLines": 1,
    "oldStart": 1,
  },
  Object {
    "changes": Array [
      Object {
        "content": "    it('App renders correctly', () => {",
        "isNormal": true,
        "newLineNumber": 21,
        "oldLineNumber": 21,
        "type": "normal",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();",
        "isDelete": true,
        "lineNumber": 22,
        "type": "delete",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();",
        "isInsert": true,
        "lineNumber": 22,
        "type": "insert",
      },
      Object {
        "content": "    });",
        "isNormal": true,
        "newLineNumber": 23,
        "oldLineNumber": 23,
        "type": "normal",
      },
    ],
    "content": "@@ -21,3 +21,3 @@ describe('basic test', () => {",
    "isPlain": false,
    "newLines": 3,
    "newStart": 21,
    "oldLines": 3,
    "oldStart": 21,
  },
]
`;

exports[`utils/diff expandFromRawCode basic 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "    it('App renders correctly', () => {",
        "isNormal": true,
        "newLineNumber": 21,
        "oldLineNumber": 21,
        "type": "normal",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();",
        "isDelete": true,
        "lineNumber": 22,
        "type": "delete",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();",
        "isInsert": true,
        "lineNumber": 22,
        "type": "insert",
      },
      Object {
        "content": "    });",
        "isNormal": true,
        "newLineNumber": 23,
        "oldLineNumber": 23,
        "type": "normal",
      },
    ],
    "content": "@@ -21,3 +21,3 @@ describe('basic test', () => {",
    "isPlain": false,
    "newLines": 3,
    "newStart": 21,
    "oldLines": 3,
    "oldStart": 21,
  },
]
`;

exports[`utils/diff textLinesToHunk basic 1`] = `
Object {
  "changes": Array [
    Object {
      "content": "",
      "isNormal": true,
      "newLineNumber": 0,
      "oldLineNumber": 0,
      "type": "normal",
    },
  ],
  "content": "@@ -0,1 +0,1",
  "isPlain": true,
  "newLines": 1,
  "newStart": 0,
  "oldLines": 1,
  "oldStart": 0,
}
`;
